File: Source.py
import math



# This is the function to encrypt the plaintext.
def encrypt(plaintxt, key):
    number_of_rows = math.ceil(len(plaintxt)/key)
    # If the plaintext length is not a multiple of the key, pad it with 'X' characters.
    if len(plaintxt) % key != 0:
        plaintxt += 'X' * (key - len(plaintxt) % key)
    encrypted_txt = ''
    # Rearranging the plaintext into a grid of characters.
    for i in range(key):
        for j in range(number_of_rows):
            if j * key + i < len(plaintxt):
                encrypted_txt += plaintxt[j * key + i]
    return encrypted_txt
# This is the function to decrypt the ciphertext.
def decrypt(ciphertxt, key):
    number_of_rows = math.ceil(len(ciphertxt)/key)
    decrypted_txt = ''
    # Rearranging the ciphertext into a grid of characters.
    for i in range(number_of_rows):
        for j in range(key):
            if i + j * number_of_rows < len(ciphertxt):
                decrypted_txt += ciphertxt[i + j * number_of_rows]
    return decrypted_txt.rstrip('X')

def main():
    user_input = input("1.Encryption\n2.Decryption\n3.Exit\n")
    if user_input == '1':
        key = input("write the Key :")
        plaintxt = input("write the plaintxt :").upper().strip().replace(" ","")
        #calling the encrypt function to encrypt the plaintext.
        encrypted_txt = encrypt(plaintxt, int(key))
        print("Encrypted text:", encrypted_txt)
    elif user_input == '2':
        key = input("write the Key :")
        ciphertxt = input("write the ciphertxt :")
        #calling the decrypt function to decrypt the ciphertext.
        decrypted_txt = decrypt(ciphertxt, int(key))
        print("Decrypted text:", decrypted_txt)
    elif user_input == '3':
        exit()
    else:
        print("Invalid option. Please try again.")
        main()

if __name__ == "__main__":
    main()

File: source.c
#include <stdio.h> // Standard input/output functions
#include <stdlib.h> // Standard library functions
#include <string.h> // String manipulation functions
#include <ctype.h> // Character type functions
#include <math.h> // Math functions

#define MAX 1000 // Maximum buffer size

// Remove spaces and convert to uppercase
void preprocess(char *s) {
    int j = 0; // Index for the processed string
    for (int i = 0; s[i]; i++) // Loop through each character in the input string
        if (!isspace(s[i])) // If the character is not a space
            s[j++] = toupper(s[i]); // Convert to uppercase and add to processed string
    s[j] = '\0'; // Null-terminate the processed string
}

// Encrypt plaintext using columnar transposition
void encrypt(char *plain, int key, char *cipher) {
    int len = strlen(plain); // Get length of plaintext
    int rows = (int)ceil((double)len / key); // Calculate number of rows
    int padded_len = rows * key; // Calculate padded length

    // Pad with 'X' if needed
    for (int i = len; i < padded_len; i++) // Loop from end of plaintext to padded length
        plain[i] = 'X'; // Pad with 'X'
    plain[padded_len] = '\0'; // Null-terminate the padded plaintext

    int idx = 0; // Index for ciphertext
    for (int col = 0; col < key; col++) // Loop through columns
        for (int row = 0; row < rows; row++) // Loop through rows
            cipher[idx++] = plain[row * key + col]; // Read column-wise and add to ciphertext
    cipher[idx] = '\0'; // Null-terminate the ciphertext
}

// Decrypt ciphertext using reverse columnar transposition
void decrypt(char *cipher, int key, char *plain) {
    int len = strlen(cipher); // Get length of ciphertext
    int rows = (int)ceil((double)len / key); // Calculate number of rows

    int idx = 0; // Index for plaintext
    for (int row = 0; row < rows; row++) // Loop through rows
        for (int col = 0; col < key; col++) { // Loop through columns
            int pos = col * rows + row; // Calculate position in ciphertext
            if (pos < len) // If position is within bounds
                plain[idx++] = cipher[pos]; // Add character to plaintext
        }
    // Remove padding
    while (idx > 0 && plain[idx - 1] == 'X') // Remove trailing 'X' padding
        idx--;
    plain[idx] = '\0'; // Null-terminate the plaintext
}

int main() {
    int choice, key; // Variables for menu choice and key
    char text[MAX], result[MAX]; // Buffers for input and output

    printf("1. Encrypt\n2. Decrypt\n3. Exit\nChoose: "); // Print menu
    scanf("%d", &choice); // Read user choice
    getchar(); // consume newline

    if (choice == 1) { // If user chooses encryption
        printf("Key: "); // Prompt for key
        scanf("%d", &key); // Read key
        getchar(); // Consume newline

        printf("Plaintext: "); // Prompt for plaintext
        fgets(text, sizeof(text), stdin); // Read plaintext
        text[strcspn(text, "\n")] = 0; // Remove newline character

        preprocess(text); // Remove spaces and convert to uppercase
        encrypt(text, key, result); // Encrypt the plaintext
        printf("Encrypted: %s\n", result); // Print encrypted text

    } else if (choice == 2) { // If user chooses decryption
        printf("Key: "); // Prompt for key
        scanf("%d", &key); // Read key
        getchar(); // Consume newline

        printf("Ciphertext: "); // Prompt for ciphertext
        fgets(text, sizeof(text), stdin); // Read ciphertext
        text[strcspn(text, "\n")] = 0; // Remove newline character

        decrypt(text, key, result); // Decrypt the ciphertext
        printf("Decrypted: %s\n", result); // Print decrypted text

    } else { // If user chooses to exit
        printf("Goodbye!\n"); // Print goodbye message
    }

    return 0; // End program
}
